### PR TITLE
Zone Lockdown Implementation

### PR DESCRIPTION
**Pull request by SteveGoldthorpe-Work**
Implementation as per:
https://github.com/terraform-providers/terraform-provider-cloudflare/issues/53

### CODE REVIEW

File: cloudflare/resource_cloudflare_zone_lockdown.go [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+
+	if urls, ok := d.GetOk("urls"); ok {
+		newZoneLockdown.URLs = expandInterfaceToStringList(urls.(*schema.Set).List())
+	}
+
+	if configurations, ok := d.GetOk("configurations"); ok {
+		newZoneLockdown.Configurations = expandZoneLockdownConfig(configurations.(*schema.Set))
+	}
+
+	log.Printf("[INFO] Creating Cloudflare Zone Lockdown from struct: %+v", newZoneLockdown)
```
[the remark this answers is not in the document]
Comment (SteveGoldthorpe-Work) [reply cloudflare/resource_cloudflare_zone_lockdown.go]: Fine, done.

File: cloudflare/resource_cloudflare_zone_lockdown.go [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+		Create: resourceCloudflareZoneLockdownCreate,
+		Read:   resourceCloudflareZoneLockdownRead,
+		Update: resourceCloudflareZoneLockdownUpdate,
+		Delete: resourceCloudflareZoneLockdownDelete,
+		Importer: &schema.ResourceImporter{
+			State: resourceCloudflareZoneLockdownImport,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"zone": {
```
[the remark this answers is not in the document]
Comment (SteveGoldthorpe-Work) [reply cloudflare/resource_cloudflare_zone_lockdown.go]: Was trying to save API calls - probably something better at the go library level if required.  Will remove zone_id and make API calls where appropriate.  Existing logic is probably flawed if zone was changed (not something I'd contemplated).

[the remark this answers is not in the document]
Comment (SteveGoldthorpe-Work) [reply cloudflare/resource_cloudflare_zone_lockdown.go]: PR has been updated to not have zone_id in the schema.

**Comment by patryk** [comment]
Looks ok, but in general I would like to support `zone_id` in all resources (look at https://github.com/terraform-providers/terraform-provider-cloudflare/blob/master/cloudflare/resource_cloudflare_access_rule.go#L100-L105).

Our intention is to refactor all resources to take `zone_id` directly as we want to provide direct zone signup functionality in the near future.

**Comment by SteveGoldthorpe-Work** [comment]
Hmm I originally had zone_id as a computed field but was asked to remove it in a previous review.  Also isn't there a race condition if we always go with a zone_id and the zone (string) changes, how would we know to re-look up the zone_id.  Should we just do that every time?

**Comment by patryk** [comment]
I've coded `access_rule` to use `zone_id` exclusivity after the initial lookup in `create` function. In schema there's `ForceNew` which will delete/create a new resource when zone/zone_id change.

The reasoning is that user can be invited to many accounts and they can have zone signed up separately (you'll end up with many different zones having the same name, obviously only one of them will be active). Doing zone lookup every time would give nondeterministic results.

Merged this one, but will refactor a bit to include the intended behaviour.

I appreciate @jacobbednarz's review here, he didn't know what we intend to move away from zone name in favour of zone ID.

**Comment by SteveGoldthorpe-Work** [comment]
Refactor - yes that's fine - hard to work in collaboration when some of the design goals aren't immediately obvious.  Is there some forum, list or chat where things like this can be discussed?  Might have a spare few cycles to help over the next few weeks.

**Comment by patryk** [comment]
Good question, we'll try to figure something out.

**Pull request merged**